src/services/cube-builder.ts

```diff
@@ -1,861 +1,1199 @@
+
+  let build: BuildLog;
+  try {
+    build = await BuildLog.startBuild(buildRevision, buildType, buildId);
+  } catch (err) {
+    logger.error(err, 'Failed to create build log entry');
+    throw err;
+  }
+
+  const createBuildSchemaRunner = dbManager.getCubeDataSource().createQueryRunner();
   try {
-    await cubeDB.query(insertQuery);
+    logger.info(`Creating schema for cube ${build.id}`);
+    await createBuildSchemaRunner.query(pgformat(`CREATE SCHEMA IF NOT EXISTS %I;`, build.id));
   } catch (error) {
-    logger.error(error, `Failed to load file into table using query ${insertQuery}`);
-    throw new FactTableValidationException(
-      'An unknown error occurred trying to load data in to the fact table.  Please contact support.',
-      FactTableValidationExceptionType.UnknownError,
-      500
-    );
+    logger.error(error, 'Something went wrong trying to create the cube schema');
+    throw error;
+  } finally {
+    void createBuildSchemaRunner.release();
+  }
+
+  let cubeBuild: CubeBuilder;
+  try {
+    cubeBuild = await createBasePostgresCube(build, dataset, buildRevision, buildType, cubeBuildConfig);
+  } catch (err) {
+    logger.error(err, 'createAllCubeFiles: Something went wrong during the actual build process');
+    if (config.cube_builder.preserve_failed) {
+      throw err;
+    }
+    const unwindBuildRunner = dbManager.getCubeDataSource().createQueryRunner();
+    try {
+      await unwindBuildRunner.query(pgformat('DROP SCHEMA IF EXISTS %I CASCADE', build.id));
+    } catch (error) {
+      logger.error(error, 'Failed to remove build schema from the data');
+    } finally {
+      createBuildSchemaRunner.release();
+    }
+    throw err;
   }
-  logger.debug(`Successfully loaded data table into fact table`);
-};
 
-export const createDatePeriodTableQuery = (factTableColumn: FactTableColumn, tableName?: string): string => {
-  if (!tableName) {
-    tableName = `${makeCubeSafeString(factTableColumn.columnName)}_lookup`;
+  if (buildType === CubeBuildType.ValidationCube) {
+    logger.debug('Validation cube build complete');
+    build.status = CubeBuildStatus.Completed;
+    build.completedAt = new Date();
+    await build.save();
+    return;
   }
-  return pgformat(
-    `
-  CREATE TABLE %I (
-    %I %s,
-    language VARCHAR(5),
-    description VARCHAR,
-    start_date TIMESTAMP WITHOUT TIME ZONE,
-    end_date TIMESTAMP WITHOUT TIME ZONE,
-    date_type varchar,
-    hierarchy %s
-  );`,
-    tableName,
-    factTableColumn.columnName,
-    factTableColumn.columnDatatype,
-    factTableColumn.columnDatatype
-  );
+
+  build.status = CubeBuildStatus.SchemaRename;
+  await build.save();
+  const createRenameSchemaRunner = dbManager.getCubeDataSource().createQueryRunner();
+  const renameStatements = [
+    'BEGIN TRANSACTION;',
+    pgformat('DROP SCHEMA IF EXISTS %I CASCADE;', buildRevision.id),
+    pgformat('ALTER SCHEMA %I RENAME TO %I;', build.id, buildRevision.id),
+    'END TRANSACTION;'
+  ];
```

Comment:
I think it would be really helpful from a db admin perspective if we prefixed at least the build schemas with `build_<build_id>` as otherwise it'll be impossible to differentiate between failed builds and actual cubes from just looking at the list of schemas without querying the build_log.